### PR TITLE
[8.x] Make MysqlDriver compatible with Doctrine\DBAL\Driver

### LINKED ISSUE #35553: Fatal Error in MysqlDriver due to incompatible method signature from interface Doctrine\DBAL\Driver
- Laravel Version: 8.18.1
- PHP Version: 7.4.12
- Database Driver & Version: MySQL 5.7.28

### Description:
We just upgraded to the latest laravel version and noticed the following fatal error in `src/Illuminate/Database/PDO/MySqlDriver.php` which gets reported from phpstan. 

```
PHP Fatal error:  Declaration of Illuminate\Database\PDO\Concerns\ConnectsToDatabase::connect(array $params) must be compatible with Doctrine\DBAL\Driver::connect(array $params, $username = NULL, $password = NULL, array $driverOptions = Array) in application/vendor/laravel/framework/src/Illuminate/Database/PDO/MySqlDriver.php on line 16
Fatal error: Declaration of Illuminate\Database\PDO\Concerns\ConnectsToDatabase::connect(array $params) must be compatible with Doctrine\DBAL\Driver::connect(array $params, $username = NULL, $password = NULL, array $driverOptions = Array) in application/vendor/laravel/framework/src/Illuminate/Database/PDO/MySqlDriver.php on line 16
```

This currently only happens in phpstan due to
```
protected function getDoctrineDriver()
    {
        return class_exists(Version::class) ? new DoctrineDriver : new MySqlDriver;
    }
```
as it loads the `MysqlDriver`-class.

### PR DESCRIPTION
**Pull request by duxthefux**
Fixes #35553

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
What version of DBAL?

File: src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php [hunk 1]
```diff
@@ -13,7 +13,7 @@ trait ConnectsToDatabase
      * @param  array  $params
      * @return \Illuminate\Database\PDO\Connection
      */
-    public function connect(array $params)
+    public function connect(array $params, $username = null, $password = null, array $driverOptions = [])
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php:16]: This is a BC break. Also, missing phpdoc.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by duxthefux** [comment]
@GrahamCampbell DBAL version 2.12.1 is the currently installed one. As far as I can tell it was caused by this commit 0942a130 which introduced support for DBAL v3.0.

This commit also added the class `src/Illuminate/Database/PDO/MySqlDriver.php` which is now causing problems due to incompatibility between the `Driver`-interface from DBAL and `src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php`-trait.

As far as I can tell it would make sense to align the method signatures from the `src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php`-trait and the `Driver`-interface from DBAL as this incompatible. Also the `getName()` method is missing.

If I'm missing something or am on the wrong path please feel free to point me to the right direction to fix it in the desired way.

Thanks for your help!

File: src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php (hunk 1 above)
Comment (duxthefux) [reply src/Illuminate/Database/PDO/Concerns/ConnectsToDatabase.php:16]: thanks for pointing out. I will fix the code up once we decided on what the desired way of solving this is. Please see my comment.

**Comment by GrahamCampbell** [comment]
So, just to clarify, everything works with any of the recent 2.x versions, but not with 3.0.0?

**Comment by duxthefux** [comment]
> So, just to clarify, everything works with any of the recent 2.x versions, but not with 3.0.0?

@GrahamCampbell yes, just tried it out!

**Comment by GrahamCampbell** [comment]
Which of the following versions works correctly:

* `2.9.3`
* `2.10.4`
* `2.11.3`
* `2.12.1`

**Comment by GrahamCampbell** [comment]
And what version(s) of PHP did you try? Only `2.12.1` would be expected to work on PHP 8.0, but the others should all work on 7.3 and 7.4.

**Comment by duxthefux** [comment]
> Which of the following versions works correctly:
> 
> * `2.9.3`
> * `2.10.4`
> * `2.11.3`
> * `2.12.1`

@GrahamCampbell 
2.9.3 => BAD
2.10.4 => BAD
2.11.3 => BAD
2.12.1 => BAD
3.x => GOOD

I currently was only able to test PHP 7.4.12
```
php -v         
PHP 7.4.12 (cli) (built: Oct 29 2020 18:37:21) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.12, Copyright (c), by Zend Technologies
```

**Comment by GrahamCampbell** [comment]
What about trying those DBAL 2.x versions with Laravel 8.14.0? Was the issue introduced in 8.15.0?

**Comment by duxthefux** [comment]
@GrahamCampbell I can confirm that it starts to break with 8.15.0 and DBAL 2.x versions

**Comment by driesvints** [comment]
@duxthefux you pointed out in your issue that this happens only when running PHPStan. Are you experiencing any problem in an actual app at all?

**Comment by duxthefux** [comment]
@driesvints yes.


You can simple reproduce it by executing this.
```
/** @var DatabaseManager $manager */
$manager = app()->make(DatabaseManager::class);
$manager->connection()->getDoctrineConnection();
```
It works fine if you have doctrine installed. But if you don't have it installed (which might be rare as some popular packages have it as a dependency) then
```
protected function getDoctrineDriver()
    {
        return class_exists(Version::class) ? new DoctrineDriver : new MySqlDriver;
    }
```
will return the `MysqlDriver` and then causes
Bildschirmfoto 2020-12-10 um 12 18 00

**Comment by driesvints** [comment]
> You can simple reproduce it by executing this.

This works fine for me on the latest 2.x DBAL version. Like I said: we don't optimize for tools like PHPStan.

**Comment by driesvints** [comment]
I'm going to close this as I don't think there's any real bug at all. If you want you can resubmit this with a failing test. Thanks

**Comment by duxthefux** [comment]
@driesvints yes, it seems like it really an edge case where in a real app this scenario is never possible and it only happens together with phpstan. 
We worked around the phpstan issue by upgrading to DBAL 3.x.

To really reproduce it changing the `class_exists(Version::class)` is necessary.

Anyway here are the steps how I managed to reproduce it.
1. Add a new test case to `tests/Database/DatabaseConnectionTest.php`
```
public function testDoctrine()
    {
        $connection = $this->getMockConnection();
        $mysqlConnection = new MySqlConnection($connection);

        // This will throw an exception
        $mysqlConnection->getDoctrineConnection();
    }
```
2. Apply patch
```
Index: src/Illuminate/Database/MySqlConnection.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/Illuminate/Database/MySqlConnection.php	(revision 3ed487766b190ef626628f71de9a8ddf2b455a80)
+++ src/Illuminate/Database/MySqlConnection.php	(date 1607610675834)
@@ -88,6 +88,6 @@
      */
     protected function getDoctrineDriver()
     {
-        return class_exists(Version::class) ? new DoctrineDriver : new MySqlDriver;
+        return new MySqlDriver;
     }
 }
```
3. Install DBAL 2.x
`composer require doctrine/dbal:^2.0`

4. Run test
```
./vendor/bin/phpunit --filter DatabaseConnectionTest::testDoctrine

PHPUnit 9.5.0 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.12
Configuration: /Applications/XAMPP/xamppfiles/htdocs/framework/phpunit.xml.dist

PHP Fatal error:  Declaration of Illuminate\Database\PDO\Concerns\ConnectsToDatabase::connect(array $params) must be compatible with Doctrine\DBAL\Driver::connect(array $params, $username = NULL, $password = NULL, array $driverOptions = Array) in /Applications/XAMPP/xamppfiles/htdocs/framework/src/Illuminate/Database/PDO/MySqlDriver.php on line 16

Fatal error: Declaration of Illuminate\Database\PDO\Concerns\ConnectsToDatabase::connect(array $params) must be compatible with Doctrine\DBAL\Driver::connect(array $params, $username = NULL, $password = NULL, array $driverOptions = Array) in /Applications/XAMPP/xamppfiles/htdocs/framework/src/Illuminate/Database/PDO/MySqlDriver.php on line 16
```.

Again this scenario never happens in runtime in a real app. Therefore closing this seems reasonable. 

Thanks for your time though!

**Pull request closed without merging**